### ISSUE TITLE
TransformerArchitecture custom layers list rejects substrate-correct architectures — relax ValidateCustomLayers to shape-based check

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1317]
## Summary

`TransformerArchitecture<T>(..., layers: List<ILayer<T>>)` runs a type-based validation that rejects any layer list whose first element isn't a "valid input layer" and whose last element isn't a "valid output layer". This blocks legitimate research architectures whose substrate at the boundaries isn't the classical dense-float-vector that AiDotNet's built-in `EmbeddingLayer<T>` / `Linear` produce.

## Repro

```csharp
var arch = new TransformerArchitecture<float>(
    inputType: InputType.TwoDimensional,
    taskType: NeuralNetworkTaskType.SequenceClassification,
    numEncoderLayers: 1, numDecoderLayers: 0, numHeads: 2,
    modelDimension: 16, feedForwardDimension: 32,
    complexity: NetworkComplexity.Medium,
    inputSize: 16, outputSize: 256,
    dropoutRate: 0.0, maxSequenceLength: 16, vocabularySize: 256,
    usePositionalEncoding: true, temperature: 1.0, sequencePooling: null,
    layers: new List<ILayer<float>> { myCustomLayer });   // any ILayer<float>

var model = new Transformer<float>(arch, new CategoricalCrossEntropyLoss<float>());
// Throws: System.ArgumentException: Invalid layer configuration:
//   The first layer must be a valid input layer.
//   The last layer must be a valid output layer.
//     at NeuralNetworkBase<T>.ValidateCustomLayersInternal(List<ILayer<T>> layers)
//     at Transformer<T>.ValidateCustomLayers(List<ILayer<T>> layers)
//     at Transformer<T>.InitializeLayers()
//     at Transformer<T>..ctor(...)
```

## Why the current behavior blocks legitimate use cases

The current `ValidateCustomLayersInternal` enforces a **type-based** contract ("first must be a specific input-layer type, last must be a specific output-layer type") when the contract that actually matters is **shape-based** ("first accepts the user-provided input shape, last produces the user-expected output shape, all layer transitions shape-compatible").

This anti-extensibility design blocks:
1. **Substrate-correct neural networks** (e.g. unit-phase manifold / U(1)^D networks, equivariant networks, neural ODEs, geometric deep learning) whose input/output layers don't match the classical `EmbeddingLayer → ... → Linear+softmax` shape. The substrate's algebra at the boundaries isn't dense float vectors with arbitrary magnitudes.
2. **Custom token-ID-to-vector bridges** that aren't `EmbeddingLayer<T>` — e.g. researcher-built embeddings with structural constraints (random complex phasors, harmonic codebooks, sparse one-hot lookups, learned positional projections).
3. **Custom readout layers** that produce probability distributions directly without a final `Linear → softmax` (e.g. retrieval-based readouts, Hopfield content-addressable readouts, Born-rule retrieval).

PyTorch's `nn.Sequential` and TensorFlow's `Sequential` both accept any sequence of modules — they validate shape compatibility, not layer-type identity. This is the right separation of concerns for an extensible framework.

## Proposed fix

Replace the type-based `ValidateCustomLayersInternal` check with one of:

**Option A — shape-based validation (preferred)**
- For the first layer: verify its `GetInputShape()` matches the user-provided input data shape
- For the last layer: verify its `GetOutputShape()` matches the loss function's expected target shape
- For every transition: verify `layers[i].GetOutputShape() == layers[i+1].GetInputShape()`
- Drop the layer-type identity check entirely

**Option B — opt-in disable flag**
- Add `disableLayerTypeValidation: bool = false` parameter to `TransformerArchitecture<T>` ctor
- When true, `ValidateCustomLayersInternal` is skipped or relaxed
- Default false preserves backward compatibility

**Option C — marker interface**
- Define `IInputLayer<T>` / `IOutputLayer<T>` marker interfaces
- Users can opt their custom layers in by implementing these markers
- `ValidateCustomLayers` checks for the interface, not the concrete type
- Library layers (`EmbeddingLayer<T>`, `DenseLayer<T>`, etc.) implement them by default

## Workaround we're using

For now we're chaining `AiModelBuilder.ConfigureLayer(ILayer<T>)` (singular) calls instead of `ConfigureModel(new Transformer<float>(arch))` — this bypasses the `Transformer<T>` constructor path that runs the strict validation. But it means we don't get `Transformer<T>`'s encoder-layer counting, attention-mask handling, auxiliary-loss support, or other Transformer-specific facilities. A relaxed validation would let us use the full Transformer pipeline with substrate-correct layer substitutions.

## Context

Building a substrate-correct (HRE / U(1)^D phase manifold) Transformer variant for head-to-head benchmarking against the classical AiDotNet Transformer. We need both arms to go through `AiModelBuilder` for fair comparison, but the input/output layer types differ at the boundary because the HRE substrate isn't classical dense-float-vector.

**Comment by ooples** [comment]
## Additional API design issue uncovered while attempting a workaround

While exploring workarounds we discovered another problem with the public surface: **`AiModelBuilder<T,TIn,TOut>.ConfigureLayer(ILayer<T> layer)` (singular) shouldn't exist in its current form.**

The fluent method name and chainable return type suggest you can build a network by chaining `ConfigureLayer(L1).ConfigureLayer(L2).ConfigureLayer(L3)...` and then calling `BuildAsync()`. But that path fails at build time with:

```
System.InvalidOperationException : Model implementation must be specified.
Use ConfigureModel() to set a model, ConfigureAutoML() for automatic model
selection, or enable agent assistance.
   at AiModelBuilder`3.BuildSupervisedInternalAsync(...)
   at AiModelBuilder`3.BuildAsync(...)
```

So `ConfigureLayer` is **non-functional unless paired with a `ConfigureModel(IFullModel)`** that already declares its layer structure. At that point, adding more layers via `ConfigureLayer` either duplicates or conflicts with the model's own layer specification. There's no clear documented semantics for what `ConfigureLayer` does in that combined state.

### Why this matters

The misleading surface area drove our entire workaround attempt: we assumed (reasonably from the fluent API shape) that `ConfigureLayer` was the "build a custom network from scratch" entry point that bypassed `ConfigureModel`'s rigid validation. Finding out it doesn't actually work that way cost a few hours of investigation.

### Suggested fix

Either:
- **Remove `ConfigureLayer(ILayer<T>)` from the public surface entirely** if it has no standalone use case. The right entry point for custom networks should be a different builder method like `ConfigureCustomNetwork(List<ILayer<T>> layers, NetworkConfig config)` that explicitly builds a generic sequential model from the layer list — bypassing any model-class-specific validation.
- **Or make it functional**: when called without a prior `ConfigureModel`, `ConfigureLayer` should accumulate layers and construct an implicit `SequentialNeuralNetwork<T>` (or equivalent) at `BuildAsync` time. This would make the chainable surface actually deliver on what its shape implies.

Either fix removes the trap. The first fix is cleaner (one less misleading method); the second is more user-friendly (the chain-fluent pattern works as expected).
